### DISCUSSION TITLE
Booting Unified Kernel Image (UKI) fails; Error registering initrd: Already started

### DISCUSSION THREAD

**Discussion by xuxuck** [ipxe/ipxe#915] [Q&A] [answered]
Hello there,
I want to boot a uki with iPXE. Fetching the image works fine but trying to boot it gives:
```
Error registering initrd: Already started
EFIIMAGE 0x5bc1eba8 could not start (or returned with error): Error 0x7f048294 (https://ipxe.org/7f048294)
Could not boot: Error 0x7f048294 (https://ipxe.org/7f048294)
```
As you can see from the error, the `ipxe.efi` I am using is build with the `DEBUG=efi_image` option and the latest version from GitHub.

Searching for `Error registering initrd: Already started` returns exactly one result namely [this](https://lore.kernel.org/lkml/20230208130835-mutt-send-email-mst@kernel.org/T/) conversation, which did not help me at all.

Booting the uki with the default PXE implementation or from disk works fine.

I would be grateful for any help.

**Comment by NiKiZe** [comment]
What is the script are you using? (To get some context)
Which version of iPXE are you using? (git commit)
If recent build, then try latest build, since some changes was made in terms of initrd handling.

The efi binary you are trying to boot from iPXE returns to with code 0x94 (or rather your firmware does) more or less described at https://ipxe.org/7f048294, or better at https://ipxe.org/7f048200, verify those debug steps to be sure)

The message "Error registering initrd: Already started" I assume actually comes from your uki.

**Reply by xuxuck** [reply]
I simply ran (manually):
```
ifopen
imgfetch http://server/uki.efi
boot uki.efi
```
The commit is 9e1f7a3659071004f4b8c76f2593da6287f0d575 so it should be the latest i guess.

Following advice from https://ipxe.org/7f048200, `imgstat` reports the correct file size and the error is the same whether secure boot is enabled or not.

**Reply by stappersg** [reply]
Thing I would like to known, is the output of
```shell
file uki.efi
```

Telling how to build, how to reproduce, the "`uki.efi` binary" would also be a good thing.

**Reply by NiKiZe** [reply]
> I simply ran (manually):
> 
> ```
> ifopen
> imgfetch http://server/uki.efi
> boot uki.efi
> ```
> 
> The commit is [9e1f7a3](https://github.com/ipxe/ipxe/commit/9e1f7a3659071004f4b8c76f2593da6287f0d575) so it should be the latest i guess.
> 
> Following advice from https://ipxe.org/7f048200, `imgstat` reports the correct file size and the error is the same whether secure boot is enabled or not.

Pefect, you can simplify that to just `chain http://server/uki.efi` or
```
kernel http://server/uki.efi
#initrd http://server/dummy
imgstat
boot
```

But I would also test with a dummy initrd, just to check that the feature in 6a004be0cceab5d669eedb5a2e6ee2feab31d5bd is not the cause of your issues. and maybe also test the parent of that commit as extra verification.

If you still have no success then maybe share more details on your uki, any linux kernel with EFI_STUB enabled should just work here.
So, share more details on how to reproduce (get or build the uki in question) would be great.

**Reply by xuxuck** [reply]
Output of `file uki.efi` is:
```
uki.efi: PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows, 12 sections
```
Maybe useful as well `objdump -f -a uki.efi`:
```
uki.efi:     file format pei-x86-64
architecture: i386:x86-64, flags 0x00000103:
HAS_RELOC, EXEC_P, D_PAGED
start address 0x0000000000003000

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00009780  0000000000003000  0000000000003000  00000400  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  1 .reloc        0000000c  000000000000d000  000000000000d000  00009c00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .data         00001fa8  000000000000e000  000000000000e000  00009e00  2**4
                  CONTENTS, ALLOC, LOAD, DATA
  3 .dynamic      00000100  0000000000010000  0000000000010000  0000be00  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  4 .rela         000004f8  0000000000011000  0000000000011000  0000c000  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .dynsym       00000018  0000000000012000  0000000000012000  0000c600  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  6 .sbat         000000e8  0000000000014000  0000000000014000  0000c800  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  7 .sdmagic      0000002e  0000000000014100  0000000000014100  0000ca00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  8 .osrel        00000163  0000000000020000  0000000000020000  0000cc00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  9 .cmdline      00000089  0000000000030000  0000000000030000  0000ce00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 10 .linux        00becc60  0000000002000000  0000000002000000  0000d000  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 11 .initrd       04daf338  0000000003000000  0000000003000000  00bf9e00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
```
The uki contains the kernel and initramfs from the Arch Linux installer iso. I built it using [sbctl](https://github.com/Foxboron/sbctl):
```
cd mountpoint/of/archlinux-2023.03.01-x86_64.iso
sbctl bundle \
-a ./arch/boot/amd-ucode.img \
-i ./arch/boot/intel-ucode.img \
-c ~/kernel-line \
-p ~/ \
-f ./arch/boot/x86_64/initramfs-linux.img \
-k ./arch/boot/x86_64/vmlinuz-linux uki.efi
```
with `~\kernel-line`:
```
archiso_nfs_srv=192.168.xxx.xxx:/volume1/pxe_media ip=dhcp archisobasedir=arch script=http://192.168.xxx.xxx/bootstrap cms_verify=y
```

**Reply by xuxuck** [reply]
> Pefect, you can simplify that to just `chain http://server/uki.efi` or
> 
> ```
> kernel http://server/uki.efi
> #initrd http://server/dummy
> imgstat
> boot
> ```

using `chain` or `kernel` and  `boot` results in the same error.

> But I would also test with a dummy initrd, just to check that the feature in [6a004be](https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd) is not the cause of your issues. and maybe also test the parent of that commit as extra verification.

Not sure what you mean by "dummy initrd" exactly so i did 
```
kernel http://server/uki.efi
initrd http://server/dummy
imgstat
boot
```
with `http://server/dummy` once beeing an empty file and once beeing the initramfs embedded inside the uki. This way after `inird http://server/dummy` it said
```
EFIIMAGE 0x43366ce8 could not load: Error 0x7f04818e (https://ipxe.org/7f04818e)
```
in both cases, whilst the first hex number `0x43366ce8` is slightly different each time. 
After `boot` the error was the same as in the beginning.

Interestingly the `EFIIMAGE [...]` error cycles 6 different colors.

**Reply by NiKiZe** [reply]
> The uki contains the kernel and initramfs from the Arch Linux installer iso. I built it using [sbctl](https://github.com/Foxboron/sbctl):
> 
> ```
> cd mountpoint/of/archlinux-2023.03.01-x86_64.iso
> sbctl bundle \
> -a ./arch/boot/amd-ucode.img \
> -i ./arch/boot/intel-ucode.img \
> -c ~/kernel-line \
> -p ~/ \
> -f ./arch/boot/x86_64/initramfs-linux.img \
> -k ./arch/boot/x86_64/vmlinuz-linux uki.efi
> ```
> 
> with `~\kernel-line`:
> 
> ```
> archiso_nfs_srv=192.168.xxx.xxx:/volume1/pxe_media ip=dhcp archisobasedir=arch script=http://192.168.xxx.xxx/bootstrap cms_verify=y
> ```

Thanks,
Maybe implement that in iPXE script instead, adding one thing at a time.
Then again build the bundle one step at a time, to see if it is any kind of bug in the bundling process, maybe that interfeers with the handles that iPXE creates for this scenario. (I understand this works in different scenarios you have tested, so it doesn't seem to be anything wrong, but it might still be a bug, so any details you can give on failing vs working might be helpful to pinpoint the issue) 

In regards to EFI debug, yes it is expected to have different colors for different instances, but really, there is little to no information that the debug output gives in this scenario, since the EFI you are trying to run "friendly returns with an error" and iPXE shows that return.

**Reply by xuxuck** [reply]
Appending the stub only with the kernel results in a kernel panic. The bundle with stub, kernel and initrd results in the same error.

I'll wait and see what the folks at systemd will say.

Thank you for your help.

**Comment by mcb30** [comment, marked as answer]
This is almost certainly a bug in the UKI stub code. Most likely it is failing to allow for the fact that there may already be an EFI handle with the Linux initrd device path installed.

What are you using to generate the UKI?

**Reply by xuxuck** [reply]
I use [sbctl](https://github.com/Foxboron/sbctl) to build the uki which in turn uses the efistub provided by systemd (version 253.1-2).

**Reply by mcb30** [reply]
Then it'll be a bug in systemd.

**Reply by xuxuck** [reply]
Ok. I submitted an issue there: https://github.com/systemd/systemd/issues/26723#issue-1616231670
Thank you for your help.

**Reply by mcb30** [reply]
Worked around in commit https://github.com/ipxe/ipxe/commit/09e8a1540.  You should now be able to boot the (still buggy) systemd-generated UKI as long as it is the only image (as shown by `imgstat`).

**Comment by LiberaVeritas** [comment]
I had the same issue, but was fixed by passing `--autofree` to the chain command
